meerk40t/core/plotplanner.py

```diff
@@ -421,35 +432,34 @@ def process(self, plot):
                 yield x, y, on
                 continue
 
-            # Shift() is on.
-            self.shift_pixels <<= 1
-            if on:
-                self.shift_pixels |= 1
-            self.shift_pixels &= 0b1111
-
-            self.shift_buffer.insert(0, (x, y))
-            if self.shift_pixels == 0b0101:
-                self.shift_pixels = 0b0011
-            elif self.shift_pixels == 0b1010:
-                self.shift_pixels = 0b1100
+            self.process_add(x, y, on)
 
+            # When buffer is full start popping off values.
             if len(self.shift_buffer) >= 4:
-                # When buffer is full start popping off values.
-                bx, by = self.shift_buffer.pop()
-                bon = (self.shift_pixels >> 3) & 1
-                yield bx, by, bon
-        # There are no more plots.
+                yield self.process_pop()
+
+    def process_add(self, x, y, on):
+        self.shift_buffer.insert(0, (x, y))
+        self.shift_pixels <<= 1
+        self.shift_pixels &= (1 << len(self.shift_buffer)) - 1
+        if on:
+            self.shift_pixels |= 1
+        assert self.shift_pixels < 16
+        if self.shift_pixels in self.PIX_SHIFT:
+            self.shift_pixels = self.PIX_SHIFT[self.shift_pixels]
```

Comment:
This code bit is rather critical and run a zillion times. There's absolutely going to be better methods of optimizing this. Please use one. Two dictionary lookups of an absolutely tiny dictionary to find 1 entry seems problematic. This could be done with `get` with a `default` value or a list-based hash-table with value->value keys. You could also possibly time the entire set to show that this code is somehow on par with methods that are almost certain to be 2x the speed at least.

Reply:
Agreed that this should be optimised.